neo4j/internal/bolt/bolt4.go

```diff
@@ -924,3 +951,24 @@ func (b *bolt4) ForceReset() error {
 	b.Reset()
 	return b.err
 }
+
+// TODO: skip JSON construction if TRACE is not enabled?
```

Comment:
Implement String interface on messages and use %s or %v as placeholder in format message, this avoids serialization if tracing is not enabled since the formatting will not be applied.

Reply:
That's a great idea.

However, that seems to only work well for `RECORD`, `SUCCESS` and `FAILURE` messages. 
The other ones are just sent as strings or maps.

An option I initially considered was to add logging to `outgoing.go` and `incoming.go` but I'd lose the Bolt version information (e.g. `appendHello` is used for both `INIT` and `HELLO`).